src/Microsoft.AspNet.OData/Query/TextPlainODataQueryOptionsParser.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using Microsoft.OData;
+using Microsoft.AspNet.OData.Common;
+using System.Diagnostics;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    public partial class TextPlainODataQueryOptionsParser
+    {
+        /// <inheritdoc/>
+        public string Parse(Stream requestStream)
+        {
+            string queryString = string.Empty;
+
+            MemoryStream memoryStream = new MemoryStream();
+            TextReader reader;
+
+            try
+            {
+                // Reset request stream position if possible
+                if (requestStream.CanSeek)
+                {
+                    requestStream.Position = 0;
+                }
+
+                requestStream.CopyTo(memoryStream);
+                memoryStream.Position = 0;
+                reader = new StreamReader(memoryStream);
+            }
+            catch
+            {
+                memoryStream.Dispose();
+
+                throw new ODataException(SRResources.CannotParseQueryOptionsPayload);
+            }
+
+            using (reader)
```

Comment:
How does this work if an exception was thrown and `reader` was never set?

Reply:
@odero The only scenario where the `reader` is not set is when an exception is thrown in the first `try...catch` and this code will never be executed in that scenario. In the `try...catch` above, we'll dispose the memory stream and exit from the function prematurely if either `CopyTo` or stream reader initialization fails. We shouldn't get to this line if we don't have an initialized reader and a memory stream